### PR TITLE
gh-112405: Optimise `pathlib.Path.relative_to`

### LINKED ISSUE #112405: Optimise `pathlib.Path.relative_to()`
# Feature or enhancement

### Proposal:

`pathlib.Path.relative_to()` can be significantly optimised by making use of `itertools.chain`

### Has this already been discussed elsewhere?

This is a minor feature, which does not need previous discussion elsewhere

### Links to previous discussion of this feature:

_No response_

### Linked PRs
* gh-112406

### PR DESCRIPTION
**Pull request by AlexWaygood**
On `main` (Windows machine, PGO-optimised build):

```
>python -m timeit -s "from pathlib import Path; p = Path('a/b/c/d/e/f/g')" "p.relative_to('a/b/c/d')"
Running PGUpdate|x64 interpreter...
10000 loops, best of 5: 28.7 usec per loop
```

With this PR:

```
>python -m timeit -s "from pathlib import Path; p = Path('a/b/c/d/e/f/g')" "p.relative_to('a/b/c/d')"
Running PGUpdate|x64 interpreter...
20000 loops, best of 5: 19.5 usec per loop
```

Fixes #112405

* Issue: gh-112405

### CODE REVIEW

**Comment by AlexWaygood** [comment]
`itertools` is an extremely fast import; this makes ~no difference to pathlib's import time :)

File: Misc/NEWS.d/next/Library/2023-11-25-20-29-28.gh-issue-112405.cOtzxC.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Optimise :meth:`pathlib.Path.relative_to`. Patch by Alex Waygood.
```
Comment (itamaro) [inline Misc/NEWS.d/next/Library/2023-11-25-20-29-28.gh-issue-112405.cOtzxC.rst]:
```suggestion
Optimize :meth:`pathlib.Path.relative_to`. Patch by Alex Waygood.
```

very nitpicky, and I don't personally mind about British vs American spelling, but "optimize" appears over 300 times in the repo, while "optimise" appears only 7 times :)

**Review by itamaro — APPROVED** [review]
nice win!

File: Misc/NEWS.d/next/Library/2023-11-25-20-29-28.gh-issue-112405.cOtzxC.rst (hunk 1 above)
Comment (AlexWaygood) [reply Misc/NEWS.d/next/Library/2023-11-25-20-29-28.gh-issue-112405.cOtzxC.rst]: Booo, both Barney and I are Brits!

File: Misc/NEWS.d/next/Library/2023-11-25-20-29-28.gh-issue-112405.cOtzxC.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Optimize :meth:`pathlib.Path.relative_to`. Patch by Alex Waygood.
```
Comment (barneygale) [inline Misc/NEWS.d/next/Library/2023-11-25-20-29-28.gh-issue-112405.cOtzxC.rst]:
```suggestion
Optimize :meth:`pathlib.PurePath.relative_to`. Patch by Alex Waygood.
```

**Review by barneygale — APPROVED** [review]
Lovely stuff

**Pull request merged**